### PR TITLE
Geonames

### PR DESCRIPTION
**Pull request by atz**
This provides needed functionality for geonames resolution, bbox extraction and ENVELOPE indexing for a new SRPT solr field.

Relates to: https://github.com/sul-dlss/exhibits/issues/209

### CODE REVIEW

File: lib/spotlight/dor/indexer.rb [hunk 1]
```diff
@@ -83,6 +91,31 @@ def add_point_bbox(sdb, solr_doc)
...
+          m = %r{^https?:\/\/sws\.geonames\.org\/(\d+)}i.match(uri.nil? ? '' : uri.value)
+          m ? m[1] : nil
+        end.reject(&:nil?).reject(&:empty?)
+      end
+
+      # Fetch remote geonames metadata and format it for Solr
+      # @param [String] id geonames identifier
+      # @return [String] Solr WKT/CQL ENVELOPE based on //geoname/bbox
+      def get_geonames_api_envelope(id)
+        url = "http://api.geonames.org/get?geonameId=#{id}&username=mejackreed" # TODO: configurable username
```
Comment (cbeer) [inline lib/spotlight/dor/indexer.rb]: This seems like an important TODO before merging, doesn't it?

File: lib/spotlight/dor/indexer.rb [hunk 2]
```diff
@@ -50,6 +52,12 @@ def add_box(sdb, solr_doc)
         solr_doc['box_ssi'] = sdb.smods_rec.box
       end
 
+      # pulls from //subject/geographic
+      def add_geonames(sdb, solr_doc)
+        ids = extract_geonames_ids(sdb)
+        solr_doc['geographic_srpt'] = ids.map { |id| get_geonames_api_envelope(id) }.reject(&:nil?)
```
Comment (cbeer) [inline lib/spotlight/dor/indexer.rb]: Does `compact` work here?

File: spec/lib/spotlight/dor/indexer_spec.rb [hunk 3]
```diff
@@ -222,38 +206,33 @@
...
-            EOF
+        let(:modsbody) do
+          <<-EOF
+            <note displayLabel="Document subtype">memorandums</note>
+            <note>a generic note</note>
+          EOF
         end
 
         it 'extracts the doc subtypes' do
-          expect(solr_doc['general_notes_ssim']).to contain_exactly 'a generic note'
```
Comment (cbeer) [inline spec/lib/spotlight/dor/indexer_spec.rb]: Why was this changed? I think rspec does better error reporting with `contain_exactly` than an ordinary `eq`, doesn't it?

File: lib/spotlight/dor/indexer.rb [hunk 4]
```diff
@@ -150,9 +183,10 @@ def add_document_subtype(sdb, solr_doc)
         solr_doc['doc_subtype_ssi'] = subtype.first unless subtype.empty?
       end
 
+      # @note upcasing each word not the same as .captialize (lowercases the rest of the string), or .titleize (breaks CamelCase words apart)
       def add_donor_tags(sdb, solr_doc)
         donor_tags = sdb.smods_rec.note.select { |n| n.displayLabel == 'Donor tags' }.map(&:content)
-        insert_field solr_doc, 'donor_tags', upcase_first_character(donor_tags), :symbol # this is a _ssim field
+        insert_field solr_doc, 'donor_tags', donor_tags.map { |v| v.sub(/^./, &:upcase) }, :symbol # this is a _ssim field
```
Comment (cbeer) [inline lib/spotlight/dor/indexer.rb]: This seems less readable to me; were you trying to address the limited utility of the `upcase_first_character` method?  Another approach might be to extract it to its own class-level helper.

Comment (atz) [inline lib/spotlight/dor/indexer.rb]: Yeah, only invoked once and was defined outside the `concerning` block.

File: spec/lib/spotlight/dor/indexer_spec.rb (hunk 3 above)
Comment (atz) [inline spec/lib/spotlight/dor/indexer_spec.rb]: For a single string in an array, no real benefit.  Value here really is expected to be an Array, not just some Enumerable (because that's what solr will accept).  I like many of the advanced rspec matchers, but I found this more accurate/readable.  If I were going to use one, it would be something like:
```ruby
expect(solr_doc).to match a_hash_including('general_notes_ssim' => ['a generic note'])
expect(solr_doc).to include('general_notes_ssim' => ['a generic note']) # or simply this
```

Because that helps when fields drift or get renamed: instead of just saying `expected "a generic note", got nil`, it shows what *is* in the hash.  And if I was going to do that, I would be changing the majority of examples in the file.

File: lib/spotlight/dor/indexer.rb [hunk 5]
```diff
@@ -150,9 +183,10 @@ def add_document_subtype(sdb, solr_doc)
         solr_doc['doc_subtype_ssi'] = subtype.first unless subtype.empty?
       end
 
+      # @note upcasing each word not the same as .captialize (lowercases the rest of the string), or .titleize (breaks CamelCase words apart)
```
Comment (mjgiarlo) [inline lib/spotlight/dor/indexer.rb]: `s/captialize/capitalize/`

File: spec/lib/spotlight/dor/indexer_spec.rb (hunk 3 above)
Comment (cbeer) [inline spec/lib/spotlight/dor/indexer_spec.rb]: I disagree. I don't think our upstream or downstream consumers care what type of thing it is, just that it contains particular elements. If this was new code, :shrug:, but changing the existing (and perfectly valid) matcher seems unnecessary.

File: lib/spotlight/dor/indexer.rb [hunk 6]
```diff
@@ -83,6 +91,31 @@ def add_point_bbox(sdb, solr_doc)
...
         solr_doc['series_ssi'] = sdb.smods_rec.series
       end
+
+      # @return [Array{String}] The IDs from geonames //subject/geographic URIs, if any
+      def extract_geonames_ids(sdb)
+        sdb.smods_rec.subject.map do |z|
+          uri = z.geographic.attr('valueURI')
+          m = %r{^https?:\/\/sws\.geonames\.org\/(\d+)}i.match(uri.nil? ? '' : uri.value)
+          m ? m[1] : nil
+        end.reject(&:nil?).reject(&:empty?)
```
Comment (cbeer) [inline lib/spotlight/dor/indexer.rb]: We're in a rails-y context here, I think? Can we use `#blank?`?

File: lib/spotlight/dor/indexer.rb [hunk 7]
```diff
@@ -83,6 +91,31 @@ def add_point_bbox(sdb, solr_doc)
       def add_series(sdb, solr_doc)
         solr_doc['series_ssi'] = sdb.smods_rec.series
       end
+
+      # @return [Array{String}] The IDs from geonames //subject/geographic URIs, if any
+      def extract_geonames_ids(sdb)
+        sdb.smods_rec.subject.map do |z|
+          uri = z.geographic.attr('valueURI')
+          m = %r{^https?:\/\/sws\.geonames\.org\/(\d+)}i.match(uri.nil? ? '' : uri.value)
```
Comment (cbeer) [inline lib/spotlight/dor/indexer.rb]: Do we need the escaped slashes? I thought `%r{}` was used to get us out of that mess?

File: spec/lib/spotlight/dor/indexer_spec.rb (hunk 3 above)
Comment (jcoyne) [inline spec/lib/spotlight/dor/indexer_spec.rb]: This change seems fine, but I think we should keep our pull requests as small as possible, which means, not changing specs unrelated to what is been changed in the code.  If you want to change the specs, you should do that in a separate pull request.

File: lib/spotlight/dor/indexer.rb [hunk 8]
```diff
@@ -83,6 +91,31 @@ def add_point_bbox(sdb, solr_doc)
...
+      def get_geonames_api_envelope(id)
+        url = "http://api.geonames.org/get?geonameId=#{id}&username=#{Spotlight::Dor::Resources::Engine.config.geonames_username}"
+        xml = Nokogiri::XML Faraday.get(url).body
+        bbox = xml.at_xpath('//geoname/bbox')
+        return if bbox.nil?
+        min_x, max_x = [bbox.at_xpath('west').text.to_f, bbox.at_xpath('east').text.to_f].minmax
+        min_y, max_y = [bbox.at_xpath('north').text.to_f, bbox.at_xpath('south').text.to_f].minmax
+        "ENVELOPE(#{min_x},#{max_x},#{max_y},#{min_y})"
+      rescue RuntimeError => e
+        logger.warn("Error fetching/parsing #{url} -- #{e.message}")
```
Comment (cbeer) [inline lib/spotlight/dor/indexer.rb]: It looks like we're not testing this exception handling at all; is it worth testing?

File: lib/spotlight/dor/indexer.rb (hunk 2 above)
Comment (atz) [inline lib/spotlight/dor/indexer.rb]: You get `map(&:value)` and `reject(&:any_other_interrogative?)`, but somehow `reject(&:nil)` is confusing?  Not feeling it.

Comment (cbeer) [inline lib/spotlight/dor/indexer.rb]: I'm not arguing it's confusing, I'm merely suggesting Ruby provides a native equivalent to `reject(&:nil?)`.

File: lib/spotlight/dor/indexer.rb (hunk 6 above)
Comment (atz) [inline lib/spotlight/dor/indexer.rb]: We're not definitively in a Rails context at all.

Comment (atz) [inline lib/spotlight/dor/indexer.rb]: The **existing** code used `&:empty?`

File: lib/spotlight/dor/indexer.rb (hunk 7 above)
Comment (atz) [inline lib/spotlight/dor/indexer.rb]: Dunno, I wrote it as a traditional `/regex/` and only `%r`'d it when rubocop complained.  Then it didn't complain.

File: lib/spotlight/dor/indexer.rb (hunk 8 above)
Comment (atz) [inline lib/spotlight/dor/indexer.rb]: It wasn't worth testing in `get_file_content`, so... In fact `get_file_content` wasn't worth testing at all, apparently.

Comment (jcoyne) [inline lib/spotlight/dor/indexer.rb]: So, write 2 tests?

File: spec/lib/spotlight/dor/indexer_spec.rb (hunk 3 above)
Comment (atz) [inline spec/lib/spotlight/dor/indexer_spec.rb]: The refactor is a separate commit, the first one.  You can see that context if you want.  If you prefer separate PRs, that's fine too.  The amount of discussion nets the same.

File: lib/spotlight/dor/indexer.rb (hunk 2 above)
Comment (atz) [inline lib/spotlight/dor/indexer.rb]: So these are ok:
- `reject(&:invalid?)`
- `reject(&:blank?)`
- `reject(&:negative?)`
- `reject(&:new?)`

But `reject(&:nil?)` is not native-enough ruby?  You can understand it perfectly well, and you can tolerate `reject` in all other cases... except this one.

File: lib/spotlight/dor/indexer.rb (hunk 7 above)
Comment (atz) [inline lib/spotlight/dor/indexer.rb]: Tested it and it seems to work fine.  Thx.

File: lib/spotlight/dor/indexer.rb (hunk 8 above)
Comment (atz) [inline lib/spotlight/dor/indexer.rb]: That would be at least three.  And rewriting several others.  Feel free to chip in here on maintenance anytime.

Comment (jcoyne) [inline lib/spotlight/dor/indexer.rb]: I expect that you will write the tests for the code that you write/change. I don't really care about `get_file_content`.

File: lib/spotlight/dor/indexer.rb (hunk 2 above)
Comment (cbeer) [inline lib/spotlight/dor/indexer.rb]: Indeed. `#compact` is the idiomatic equivalent to `reject(&:nil?)` and all of the other ways you could choose to write it.

Comment (jkeck) [inline lib/spotlight/dor/indexer.rb]: FWIW `compact` works [down at the C layer](https://github.com/ruby/ruby/blob/f14d64638a461ff660d16b42ae827b46a0cec91b/array.c#L4412), so it is more efficient:

```
Benchmark.bm do |x|
  x.report('Compact:') { array.compact }
  x.report('Reject nil:') { array.reject(&:nil?) }
end
          user     system      total        real
Compact:     0.000000   0.000000   0.000000 (  0.000008)
Reject nil:  0.000000   0.000000   0.000000 (  0.000102)
```

File: lib/spotlight/dor/indexer.rb (hunk 8 above)
Comment (jcoyne) [inline lib/spotlight/dor/indexer.rb]: This should probably be `logger.error` and I think we may want to include the back trace too.  That would be helpful in tracking down where these errors are occurring.

File: lib/spotlight/dor/indexer.rb (hunk 2 above)
Comment (atz) [inline lib/spotlight/dor/indexer.rb]: In a way that for this code literally will never matter.

File: lib/spotlight/dor/indexer.rb (hunk 8 above)
Comment (atz) [inline lib/spotlight/dor/indexer.rb]: I agree w/ the `logger.error` part anyway.

File: lib/spotlight/dor/indexer.rb (hunk 2 above)
Comment (cbeer) [inline lib/spotlight/dor/indexer.rb]: I assume from the 👍 's @jcoyne and @mjgiarlo added, they also have a preference for `#compact` here. Is `#compact` really something you just can't live with?

Comment (atz) [inline lib/spotlight/dor/indexer.rb]: Here, you are exercising your preference in a truly arbitrary way over trivialities.  This distracts from larger gestalt problems like _this gem shouldn't even exist_.  You should apply a reasonable standard of acceptability, not stylistic exactitude even **beyond rubocop**.  We do not have *exactly* the same style and we don't have to.  Where no rule is formalized, you should anticipate and allow some variation.  You understand what I wrote perfectly well.  The performance is not an actual issue.  This is below the threshold of what should even prompt discussion.  The individual changes: Yeah, I can live with that.  The arbitrary exactitude: No, I can't.

**Comment by cbeer** [comment]
👍  can you squash the last two commits, though?

**Comment by atz** [comment]
Nah.
